timemachine/cpp/src/nonbonded_mol_energy.cu

```diff
@@ -0,0 +1,112 @@
+#include "gpu_utils.cuh"
+#include "kernels/k_nonbonded.cuh"
+#include "kernels/kernel_utils.cuh"
+#include "math_utils.cuh"
+#include "mol_utils.hpp"
+#include "nonbonded_mol_energy.hpp"
+
+const static int BLOCK_SIZE = 256;       // 256 blocks appears fastest
+const static int THREADS_PER_BLOCK = 64; // 64 threads seem to produce the best outcome on Cuda Arch 8.6
+
+namespace timemachine {
+
+template <typename RealType>
+NonbondedMolEnergyPotential<RealType>::NonbondedMolEnergyPotential(
+    const int N, const std::vector<std::vector<int>> target_mols, const double beta, const double cutoff)
+    : N_(N), num_target_mols_(target_mols.size()), beta_(static_cast<RealType>(beta)),
+      cutoff_squared_(static_cast<RealType>(cutoff * cutoff)) {
+    verify_group_idxs(N_, target_mols);
+
+    std::array<std::vector<int>, 3> target_flattened_groups = prepare_group_idxs_for_gpu(target_mols);
+
+    d_target_atom_idxs_.realloc(target_flattened_groups[0].size());
```

Comment:
would be nice if we could set these members to `const`, (probably time for another constructor that takes in a vector)

Reply:
(non-blocking, can save for deboggling)

Reply:
We would need to decompose the target group indices prior to the constructor to do that. Which we could do in the wrapped kernels, though not sure how much of that sort of logic in the wrapped kernels.